instance/models/server.py

```diff
@@ -0,0 +1,274 @@
+            return None
+
+        return public_addr['addr']
+
+    def update_status(self, provisioned=False, rebooting=False):
+        """
+        Refresh the status by querying the openstack server via nova
+        """
+        # TODO: Check when server is stopped or terminated
+        os_server = self.os_server
+        self.log('debug', 'Updating status for {} from nova (currently {}):\n{}'.format(
+            self, self.status, to_json(os_server)))
+
+        if self.status == 'started':
+            self.log('debug', 'Server {}: loaded="{}" status="{}"'.format(
+                self, os_server._loaded, os_server.status))
+            if os_server._loaded and os_server.status == 'ACTIVE':
+                self._set_status('active')
+
+        elif self.status == 'active' and is_port_open(self.public_ip, 22):
+            self._set_status('booted')
+
+        elif self.status == 'booted' and provisioned:
+            self._set_status('provisioned')
+
+        elif self.status in ('provisioned', 'ready') and rebooting:
+            self._set_status('rebooting')
+
+        elif self.status == 'rebooting' and not rebooting and is_port_open(self.public_ip, 22):
+            self._set_status('ready')
+
+        return self.status
+
+    def start(self):
+        """
+        Get a server instance started and an openstack_id assigned
+
+        TODO: Add handling of quota limitations & waiting list
+        TODO: Create the key dynamically
+        """
+        self.log('info', 'Starting server {} (status={})...'.format(self, self.status))
+        if self.status == 'new':
+            os_server = openstack.create_server(
+                self.nova,
+                self.instance.sub_domain,
+                settings.OPENSTACK_SANDBOX_FLAVOR,
+                settings.OPENSTACK_SANDBOX_BASE_IMAGE,
+                key_name=settings.OPENSTACK_SANDBOX_SSH_KEYNAME,
+            )
+            self.openstack_id = os_server.id
+            self.log('info', 'Server {} got assigned OpenStack id {}'.format(self, self.openstack_id))
+            self._set_status('started')
+        else:
+            raise NotImplementedError
+
+    def reboot(self, reboot_type='SOFT'):
+        """
+        Reboot the server
+
+        This requires to switch the status to 'rebooting', which is first attempted via the
+        `update_status` method. If the current state doesn't allow to switch to this status,
+        a ServerNotReady exception is thrown.
+        """
+        if self.update_status(rebooting=True) != 'rebooting':
+            raise ServerNotReady("Can't change status to 'rebooting' (current: '{}')".format(self.status))
+        self.os_server.reboot(reboot_type=reboot_type)
+
+        # TODO: Find a better way to wait for the server shutdown and reboot
+        # Currently, without sleeping here, the status would immediately switch back to ready,
+        # as SSH is still available until the reboot terminates the SSHD process
```

Comment:
I don't think there really is one. Some people recommend using cloud-init **phone home** http://cloudinit.readthedocs.org/en/latest/topics/examples.html#call-a-url-when-finished but that requires that there is a publicly available IP to do that. I chose to have a function dealing with all kind of ssh cnx errors and looking for cloud-init termination string in the log files instead, but I'm not entirely happy about it https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/__init__.py#L167 and the user-data file that reboots during cloud-init https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/openstack-centos-6.5-user-data.txt